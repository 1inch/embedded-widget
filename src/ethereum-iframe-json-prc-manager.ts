export interface EthereumRequestArguments {
    id: string | number;
    method: string;
    params?: any;
}

export interface EthereumProvider {
    request(args: EthereumRequestArguments): Promise<any>;
    enable(args: EthereumRequestArguments): Promise<any>;
}

export class EthereumIframeJsonRpcManager {
    private readonly eventsListener: (e: MessageEvent) => void;

    private onIframeLoadCallback: (() => void) | null = null;

    private isIframeLoaded = false;

    constructor(
        private iframe: HTMLIFrameElement,
        private ethereumProvider: EthereumProvider
    ) {
        this.eventsListener = event => {
            if (event.data.jsonrpc === '2.0') {
                this.processRequest(event.data);
            }
        };

        window.addEventListener('message', this.eventsListener);

        this.iframe.addEventListener('load', () => {
            this.onIframeLoadCallback?.();
            this.isIframeLoaded = true;
        });
    }

    onIframeLoad(callback: () => void) {
        if (this.isIframeLoaded) {
            callback();
            return;
        }

        this.onIframeLoadCallback = callback;
    }

    destroy() {
        window.removeEventListener('message', this.eventsListener);

        this.iframe.parentElement?.removeChild(this.iframe);
    }

    processRequest(request: EthereumRequestArguments): Promise<void> {
        if (!this.ethereumProvider) {
            return Promise.resolve();
        }

        const request$ = request.method === 'enable'
            ? this.ethereumProvider.enable(request)
            : this.ethereumProvider.request(request);

        const contentWindow = this.iframe.contentWindow!;

        return request$.then(result => {
            contentWindow.postMessage({
                jsonrpc: '2.0',
                id: request.id,
                result
            }, '*');
        }).catch(error => {
            contentWindow.postMessage({
                jsonrpc: '2.0',
                id: request.id,
                error
            }, '*');
        });
    }
}
