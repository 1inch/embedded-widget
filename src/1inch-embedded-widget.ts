import {EthereumIframeJsonRpcManager, EthereumProvider} from './ethereum-iframe-json-prc-manager';

const defaultOneInchOrigin = 'https://app.1inch.io';
const iframeId = 'oneInchWidgetFrame';

export interface OneInchWidgetOptions {
    chainId: number;
    sourceTokenSymbol: string;
    destinationTokenSymbol: string;
    hostElement: HTMLElement;
    provider: EthereumProvider;
    theme?: 'light' | 'dark';
    oneInchOrigin?: string;
    sourceTokenAmount?: string;
}

export function setup1inchWidget(options: OneInchWidgetOptions): EthereumIframeJsonRpcManager {
    const {
        hostElement,
        provider,
        chainId,
        sourceTokenAmount,
        sourceTokenSymbol,
        destinationTokenSymbol,
        oneInchOrigin,
        theme
    } = options;

    const iframe = document.createElement('iframe');
    const query = new URLSearchParams({
        sourceTokenAmount: sourceTokenAmount || '',
        theme: theme || ''
    });
    const origin = oneInchOrigin || defaultOneInchOrigin;
    const existedFrame = document.getElementById(iframeId);

    iframe.id = iframeId;
    iframe.src = `${origin}/#/${chainId}/embedded-swap/${sourceTokenSymbol}/${destinationTokenSymbol}?${query}`;
    if (existedFrame) {
        hostElement.removeChild(existedFrame);
    }
    hostElement.appendChild(iframe);

    return new EthereumIframeJsonRpcManager(iframe, provider);
}
