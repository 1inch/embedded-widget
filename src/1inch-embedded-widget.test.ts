import {OneInchWidgetOptions, setup1inchWidget} from './1inch-embedded-widget';
import {EthereumIframeJsonRpcManager, EthereumProvider} from './ethereum-iframe-json-prc-manager';
import {instance, mock} from 'ts-mockito';

describe('setup1inchWidget', () => {
    let ethereumProvider: EthereumProvider;

    beforeEach(() => {
        ethereumProvider = mock<EthereumProvider>();
    });

    it('Should setup iframe with correct url', () => {
        const host = document.createElement('div');

        const options: OneInchWidgetOptions = {
            chainId: 137,
            sourceTokenSymbol: '1INCH',
            destinationTokenSymbol: 'DAI',
            hostElement: host,
            provider: instance(ethereumProvider),
            theme: 'light',
            sourceTokenAmount: '15'
        };

        const rpcManager = setup1inchWidget(options);

        expect(rpcManager).toBeInstanceOf(EthereumIframeJsonRpcManager);
        expect(host.innerHTML).toMatchSnapshot();
    });
});
