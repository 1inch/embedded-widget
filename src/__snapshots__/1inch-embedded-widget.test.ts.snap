// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`setup1inchWidget Should setup iframe with correct url 1`] = `"<iframe id=\\"oneInchWidgetFrame\\" src=\\"https://app.1inch.io/#/137/embedded-swap/1INCH/DAI?sourceTokenAmount=15&amp;theme=light\\"></iframe>"`;
